external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -169,16 +169,16 @@ private[sql] class AvroDeserializer(
         }
         updater.set(ordinal, bytes)
 
-      case (FIXED, d: DecimalType) => (updater, ordinal, value) =>
-        val bigDecimal = decimalConversions.fromFixed(value.asInstanceOf[GenericFixed], avroType,
-          LogicalTypes.decimal(d.precision, d.scale))
-        val decimal = createDecimal(bigDecimal, d.precision, d.scale)
+      case (FIXED, _: DecimalType) => (updater, ordinal, value) =>
+        val d = avroType.getLogicalType.asInstanceOf[LogicalTypes.Decimal]
+        val bigDecimal = decimalConversions.fromFixed(value.asInstanceOf[GenericFixed], avroType, d)
+        val decimal = createDecimal(bigDecimal, d.getPrecision, d.getScale)
```

Comment:
So here we create the Decimal of precision/scale for the logical type of avro, not for Catalyst type?

Reply:
Yes, it should be because `createDecimal` do `Decimal(decimal.unscaledValue().longValue(), precision, scale)`.

Reply:
Will we cast the decimal value to the catalyst decimal type precision and scale? It looks risky if the value inside `InternalRow` doesn't match the data type.

I'm also curious about behaviors when the precision from avro file schema is much larger than the catalyst decimal tyoe, do we truncate the value?

Reply:
For now, we don't cast here yet. Are you suggesting it?
> Will we cast the decimal value to the catalyst decimal type precision and scale? It looks risky if the value inside InternalRow doesn't match the data type.

For the following, could you give me some example? `much larger` is ambiguous to me. I can take a look.
> I'm also curious about behaviors when the precision from avro file schema is much larger than the catalyst decimal tyoe, do we truncate the value?

In general, the following is the expected behavior for decimal range. 
```scala
scala> spark.read.format("avro").load("/tmp/avro").show
+----+
|   a|
+----+
|3.14|
+----+


scala> spark.read.format("avro").load("/tmp/avro").printSchema
root
 |-- a: decimal(3,2) (nullable = true)


scala> spark.read.schema("a DECIMAL(2, 1)").format("avro").load("/tmp/avro").show()
+---+
|  a|
+---+
|3.1|
+---+


scala> spark.read.schema("a DECIMAL(4, 3)").format("avro").load("/tmp/avro").show()
+-----+
|    a|
+-----+
|3.140|
+-----+
```

Reply:
BTW, @cloud-fan . I want to make it sure the following.
1. This PR have no side-effect when the physical schema matches the catalyst schema.
2. This PR fixes the incorrect result when the physical schema mismatches the catalyst schema.

This is much better than before. And, we can fix the inconsistency in the upper-layer still because we have the original decimal's precision and scale still.

Reply:
```
scala> spark.read.schema("a DECIMAL(2, 1)").format("avro").load("/tmp/avro").show()
+---+
|  a|
+---+
|3.1|
+---+
```

This answers my question. So truncation happens, but I don't know where it happens yet.

Yea it's better than before, I just want to make sure there are no more correctness bugs here.